### DISCUSSION TITLE
Support for HMC5883L 3axis Compass

### DISCUSSION THREAD

**Discussion by elik745i** [arendst/Tasmota#10558] [Ideas]
I'm using GY-87 (GY-80) in the Weather Station project. GY-87 has MPU-6050, BMP and HMC5883L sensors on it. I can get data from all installed sensors but I need Compass data to get direction of the wind. 
I have found old libraries here:
[https://git.koehlerweb.org/frodovdr/Sonoff-Tasmota/commit/7f567ccf217bb7867b305f85b3dd02f1696b21a9](url)
, but those does not work. I have tuned them to compile in Visual Studio:
[I2Cdevlib-HMC5883L.zip](https://github.com/arendst/Tasmota/files/5814857/I2Cdevlib-HMC5883L.zip)
[xsns_81_hmc5883l.zip](https://github.com/arendst/Tasmota/files/5814859/xsns_81_hmc5883l.zip)

 but still sensor does not appear in the Tasmota page. Please help!

**Comment by barbudor** [comment]
Without the device, we can't debug that for you
You need to add logs in the drivers in suitable places to find if it is detected, if the functions are called etc ...

**Reply by elik745i** [reply]
> Without the device, we can't debug that for you
> You need to add logs in the drivers in suitable places to find if it is detected, if the functions are called etc ...

Thanks for reply, maybe you can suggest existing driver, which could be easier to modify for that device? I look at MPU-6050 and trying to do the same, but that is kinda too complicated..

**Reply by elik745i** [reply]
I'm modifying *.ino cpp and*.h files, also added line in I2C DEVICES.md file:

  56  | USE_HMC5883L        | xsns_81 | HMC5883L | 0x1E        | Magnetic Field Sensor

What else need to be modified in tasmota source folder?

I got a feeling that I miss something that is why device is not detected..

**Reply by barbudor** [reply]
What files of Tasmota (apart from the above) have you modified ?
Normally you shouldn't have changed any of the Tasmota core files

Note that if you are using the latest development fom Github, XNS_81 and XI2C_56 are already in use
There is a risk of conflict
I suggest you use XSNS_99 and XI2C_96. It can be changed later if your changes get integrated.

Try to have only your sensors in your build to avoid any conflict.

If you type `i2cscan` at the console, is the HM detected ?

**Reply by barbudor** [reply]
For better readability, remove the paste of the code in the 1st post and attach the files .

**Reply by barbudor** [reply]
The discussion in https://github.com/arendst/Tasmota/pull/5165 make it clear that the proposed code is not complete, missing parts such as calibration.
Are you really wanting to spend time with a code that is known to not be completed and that doesn't seem to be able to maintain yourself ?

**Reply by elik745i** [reply]
OK will use suggested XSNS_99 and XI2C_96, nothing else is modified.

**Reply by elik745i** [reply]
15:04:20 CMD: i2cscan
15:04:20 MQT: stat/WStation/RESULT = {"I2CScan":"Device(s) found at 0x1e 0x53 0x69 0x77"}

**Reply by elik745i** [reply]
My device is at 0x1E

**Comment by elik745i** [comment]
[HMC5883L_3-Axis_Digital_Compass_IC.pdf](https://github.com/arendst/Tasmota/files/5814842/HMC5883L_3-Axis_Digital_Compass_IC.pdf)

I'm attaching manual for HMC5883L maybe that can give you idea. For calibration, I need at least to have readings of the sensor to be shown, then I can think of callibration..

**Reply by elik745i** [reply]
Thanks for help again!

**Reply by barbudor** [reply]
Does the driver detect the HMC and reports values now ?
Do not hesitate to use function `AddLog_P(  .. )` in your code to add logs in the console.
There are plenty of example in other drivers how to use that function.

**Reply by elik745i** [reply]
> Does the driver detect the HMC and reports values now ?

Nop, all it reports:
``
15:11:03 MQT: tele/WStation/SENSOR = {"Time":"2021-01-14T15:11:03","COUNTER":{"C1":0},"ANALOG":{"A0":1024},"DHT11":{"Temperature":null,"Humidity":null,"DewPoint":null},"BMP180":{"Temperature":23.9,"Pressure":1003.1},"PressureUnit":"hPa","TempUnit":"C"}
``

> Do not hesitate to use function `AddLog_P( .. )`    --  will figure it out how, thanks

**Reply by elik745i** [reply]
Strange I have added XSNS100 as changing to XNSN99 gives error while compilling, same did 100:

``
E:/PROJECTS/TASMOTA/0.92/Tasmota-master/tasmota/xsns_interface.ino:423:4: error: 'Xsns100' was not declared in this scope
   &Xsns100
    ^
*** [.pio\build\tasmota\src\tasmota.ino.cpp.o] Error 1
``

**Reply by elik745i** [reply]
WHY? Where it should be declared in which file?

**Reply by elik745i** [reply]
Found it at the end of the file, thanks to @vic

**Reply by barbudor** [reply]
Sent you also in Discord to read https://tasmota.github.io/docs/Sensor-API/

**Reply by elik745i** [reply]
Can someone look at the driver I write, Compilling OK, but Tasmota gives error log ((
[xsns_99_hmc5883l.zip](https://github.com/arendst/Tasmota/files/5817136/xsns_99_hmc5883l.zip)

**Comment by elik745i** [comment]
Thanks for help. I got a bit of progress. I got it to show and detect Magnetometer, but! showings are 0 (((
Cas some GURU have a look to the manual attached above and the latest code written by me:
[xsns_99_hmc5883l.zip](https://github.com/arendst/Tasmota/files/5817179/xsns_99_hmc5883l.zip)
and suggest/fix?

**Comment by elik745i** [comment]
![HMC5883L](https://user-images.githubusercontent.com/17794736/104750578-bb475c80-576d-11eb-8cfc-a5acc927b885.png)
HeyHoo!! I did it!
attached is the working driver for compass.
Few more tasmota filess need to be changed (added definitions and units) to get it working.
This is my first driver for tasmota, so do not crucify me for not using libraries and putting everything into one file.
Thanks everyone for help!
[xsns_99_hmc5883l.zip](https://github.com/arendst/Tasmota/files/5821626/xsns_99_hmc5883l.zip)

**Comment by elik745i** [comment]
To get HMC chip to be identified and working on GY-87 board need to add into initialization (beginning) the following:

`//Bypass Mode to get HMC module on GY-87 working!
Wire.beginTransmission(0x68);
Wire.write(0x37);
Wire.write(0x02);
Wire.endTransmission();

Wire.beginTransmission(0x68);
Wire.write(0x6A);
Wire.write(0x00);
Wire.endTransmission();

//Disable Sleep Mode
Wire.beginTransmission(0x68);
Wire.write(0x6B);
Wire.write(0x00);
Wire.endTransmission();
//Bypass Mode to get HMC module on GY-87 working!`

**Comment by elik745i** [comment]
Updated version of the driver, which includes fix for working on GY-87 board.
[xsns_99_hmc5883l.zip](https://github.com/arendst/Tasmota/files/5824639/xsns_99_hmc5883l.zip)

**Reply by Bolukan** [reply]
Thx for this example, as I intend to rewrite it for a QMC5883L!

**Reply by fir3drag0n** [reply]
Do you have a working version for QMC5883L, in particular GY-271?

**Reply by sfromis** [reply]
Ideally, the suggested driver code could be submitted as a PR for review/inclusion in standard Tasmota, as an optional feature.

**Comment by elik745i** [comment]
Compiled Firmware and Flasher:
[firmware.zip](https://github.com/arendst/Tasmota/files/5826124/firmware.zip)

**Reply by Sparal** [reply]
Hi, 
I was hoping for your firmware to work, but I cannot get the sensor to send some values.
On the D1 Mini I have the following config:
Modul type: Generic (18)
D2: I2C SCL
D1: I2C SDA
Seems the firmware don't recognize the sensor, because i get no value on the main page. 
But the command "i2cscan" return: {"I2CScan":"Device(s) found at 0x0d"}
Do I do something wrong?

**Reply by sfromis** [reply]
You may be interested in a current thread about support for the sensor. #16705

**Comment by iot-sle** [comment]
Hi, I guess you finished and debugged your device driver now. Seems to be a nice pice of software, but you need to generate a pull request to get it part of the Tasmota project. I  'm searching for just that device driver but I didn't find it in Tasmota sources now. Do you plan to do the pull request?

**Reply by elik745i** [reply]
I'm too lazy to find how to do that? please somebody do it for me, all required files are here.

**Comment by iot-sle** [comment]
If Tasmota admins are ok with that, I can take a look after Easter. Do some tests and if it’s working do the pull request.

 

 

Von: elik745i ***@***.*** 
Gesendet: Dienstag, 30. März 2021 18:46
An: arendst/Tasmota
Cc: iot-sle; Manual
Betreff: Re: [arendst/Tasmota] Support for HMC5883L 3axis Compass (#10558)

 

I'm too lazy to find how to do that? please somebody do it for me, all required files are here.

**Reply by ascillato** [reply]
All pull requests are very welcome

**Comment by BoBeRzE** [comment]
Hi,

any news about the driver? it would be great if this devis is working with tasmota.

BoBeRzE

**Comment by KNXfriend** [comment]
Hello all together,
manny thanks for this great work!
I am as well interested in this compass support. What can I do, to support this integration?
So far I not really understand tho whole compile process. What I need, is a combination of BMP280 and a magnetic compass module.
Any news or jobs to do?
